Override https://repo.gradle.org/gradle/list/libs-releases via Gradle properties (with optional credentials)

kelemen opened the pull request:
Fixes #25840. This PR will let users to change the <https://repo.gradle.org/gradle/list/libs-releases> with the `org.gradle.libraries.sourceRepository.url` Gradle property (with a fallback to `GRADLE_LIBS_REPO_OVERRIDE` environment variable). In addition to the URL, it is possible to override the credentials via the `org.gradle.libraries.sourceRepository.credentialsProperties` Gradle property which may contain the Gradle properties pointing to the username and password.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

ov7a commented:
Thank you for your contribution!

This PR has a valid DCO and tests. The relevant team for this area will confirm the design of the implementation choices.

***

FTR, there is a related issue about potential solution to this 
* https://github.com/gradle/gradle/issues/29049

Previous PRs were closed. @gradle/bt-jvm given the history, can you decide if #25840 should be closed in favor of #29049

kelemen commented:
Is there a very likely ETA for #29049? Because if there is (and it is not years) then that would obviously be a superior solution, but if there is no clear date, then until then this would be a good temporary workaround. Though I was now thinking that the way I implemented this was a mistake, but if this would only be temporary, then the way I implemented it might even be better, because people could harmlessly commit both Gradle properties into their repository (given that it does not contain sensitive information), and the properties would remain harmless even after the eventual real fix.

tresat commented:
@kelemen - There is not an ETA for that issue.  This PR is queued for triaging, we will consider it as a workaround then.

big-guy commented:
Sorry, we don't want to make this configurable via system properties.  We're working on making the localGroovy and gradleApi dependencies behave more like real dependencies, which will then just work with the project's defined repositories.  

Until then, we might consider an option to disable this entirely.  We don't want more complexity here with how we find sources for these dependencies.

kelemen commented:
@big-guy Understandable. Though just for the note, I did it with Gradle properties, not system properties.

That said, the complexity is already there since you have to set an environment variable, which is the most annoying way to set it because they stick with the already started processes, making it quite painful to keep troubleshooting people who don't realize this.

So, even if you don't accept this temporary workaround, please find something else than the current environment variable based in the near future.
